File: Lynx-IOS/Lynx-IOS/Networking/APIClient.swift
import Foundation

enum APIError: Error {
    case http(Int)
    case notFound
    case decoding
}

/// REST client for the Lynx backend (see backend/src/index.ts).
/// A native app is not subject to CORS, so this talks to Cloud Run directly.
struct APIClient {
    let accessToken: String

    private var decoder: JSONDecoder { JSONDecoder() }

    private func authorizedRequest(_ path: String, method: String = "GET") -> URLRequest {
        var request = URLRequest(url: AppConfig.apiBaseURL.appending(path: path))
        request.httpMethod = method
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        return request
    }

    func listConversations() async throws -> [Conversation] {
        let (data, response) = try await URLSession.shared.data(for: authorizedRequest("/v1/conversations"))
        try Self.checkStatus(response)
        let decoded = try decoder.decode(ConversationListResponse.self, from: data)
        return decoded.data.map(Conversation.init)
    }

    func createConversation() async throws -> Conversation {
        let (data, response) = try await URLSession.shared.data(
            for: authorizedRequest("/v1/conversations", method: "POST")
        )
        try Self.checkStatus(response)
        let server = try decoder.decode(ServerConversation.self, from: data)
        return Conversation(server: server)
    }

    struct MessagePage {
        let messages: [ChatMessage]
        let nextCursor: String?
        let hasMore: Bool
    }

    func getMessages(conversationId: String, limit: Int = 50, before: String? = nil) async throws -> MessagePage {
        var components = URLComponents(
            url: AppConfig.apiBaseURL.appending(path: "/v1/conversations/\(conversationId)/messages"),
            resolvingAgainstBaseURL: false
        )!
        var query = [URLQueryItem(name: "limit", value: String(limit))]
        if let before { query.append(URLQueryItem(name: "before", value: before)) }
        components.queryItems = query

        var request = URLRequest(url: components.url!)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)
        try Self.checkStatus(response)
        let decoded = try decoder.decode(MessagesPageResponse.self, from: data)
        return MessagePage(
            messages: decoded.data.map(ChatMessage.init),
            nextCursor: decoded.nextCursor,
            hasMore: decoded.hasMore
        )
    }

    private static func checkStatus(_ response: URLResponse) throws {
        guard let http = response as? HTTPURLResponse else { return }
        if http.statusCode == 404 { throw APIError.notFound }
        guard (200..<300).contains(http.statusCode) else { throw APIError.http(http.statusCode) }
    }
}

File: Lynx-Andrioid/app/src/main/java/com/anantha/lynx/network/ApiClient.kt
package com.anantha.lynx.network

import com.anantha.lynx.model.ChatMessage
import com.anantha.lynx.model.Conversation
import com.anantha.lynx.model.ConversationListResponse
import com.anantha.lynx.model.MessagesPageResponse
import com.anantha.lynx.model.ServerConversation
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.serialization.json.Json
import okhttp3.Call
import okhttp3.Callback
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.IOException
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

sealed class ApiError : Exception() {
    data class Http(val code: Int) : ApiError()
    object NotFound : ApiError()
    data class Decoding(override val cause: Throwable) : ApiError()
}

/**
 * REST client for the Lynx backend (see backend/src/index.ts). A native app
 * is not subject to CORS, so this talks to Cloud Run directly.
 * Ported from Lynx-IOS/Lynx-IOS/Networking/APIClient.swift.
 */
class ApiClient(private val accessToken: String) {

    private val json = Json { ignoreUnknownKeys = true }

    private fun authorizedRequest(path: String, method: String = "GET"): Request.Builder =
        Request.Builder()
            .url(AppConfig.API_BASE_URL.toHttpUrl().newBuilder().addPathSegments(path.trimStart('/')).build())
            .method(method, null)
            .addHeader("Authorization", "Bearer $accessToken")

    suspend fun listConversations(): List<Conversation> {
        val response = execute(authorizedRequest("v1/conversations").build())
        val decoded = decode<ConversationListResponse>(response)
        return decoded.data.map(::Conversation)
    }

    suspend fun createConversation(): Conversation {
        val response = execute(authorizedRequest("v1/conversations", "POST").build())
        val decoded = decode<ServerConversation>(response)
        return Conversation(decoded)
    }

    data class MessagePage(
        val messages: List<ChatMessage>,
        val nextCursor: String?,
        val hasMore: Boolean,
    )

    suspend fun getMessages(conversationId: String, limit: Int = 50, before: String? = null): MessagePage {
        val urlBuilder = AppConfig.API_BASE_URL.toHttpUrl().newBuilder()
            .addPathSegments("v1/conversations/$conversationId/messages")
            .addQueryParameter("limit", limit.toString())
        if (before != null) urlBuilder.addQueryParameter("before", before)

        val request = Request.Builder()
            .url(urlBuilder.build())
            .addHeader("Authorization", "Bearer $accessToken")
            .build()

        val response = execute(request)
        val decoded = decode<MessagesPageResponse>(response)
        return MessagePage(
            messages = decoded.data.map(::ChatMessage),
            nextCursor = decoded.nextCursor,
            hasMore = decoded.hasMore,
        )
    }

    private suspend fun execute(request: Request): Response {
        val response = sharedHttpClient.newCall(request).await()
        if (response.code == 404) throw ApiError.NotFound
        if (response.code !in 200..299) throw ApiError.Http(response.code)
        return response
    }

    private inline fun <reified T> decode(response: Response): T {
        val body = response.use { it.body.string() }
        return try {
            json.decodeFromString(body)
        } catch (e: Exception) {
            throw ApiError.Decoding(e)
        }
    }

    companion object {
        val sharedHttpClient: OkHttpClient = OkHttpClient.Builder().build()
    }
}

suspend fun Call.await(): Response = suspendCancellableCoroutine { continuation ->
    enqueue(object : Callback {
        override fun onFailure(call: Call, e: IOException) {
            continuation.resumeWithException(e)
        }

        override fun onResponse(call: Call, response: Response) {
            continuation.resume(response)
        }
    })
    continuation.invokeOnCancellation { cancel() }
}
